### PR TITLE
Lua: consolidate setup hooks

### PR DESCRIPTION
**Pull request by Shados**
###### Motivation for this change
- Means there are no longer two separate setup hooks for Lua packages, with partially-overlapping functionality.
- Fixes pre-existing issues that came to light [here](https://github.com/NixOS/nixpkgs/pull/63108#issuecomment-508670438).

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @teto @vcunat can one of you two poke ofborg for this? I know from last time it would take me a couple days to do a nix-review for all the builds >.>.

Both prosody tests pass, I've manually tested luakit, and I've tested a handful of `buildLuaPackage` and `buildLuarocksPackage` derivations, all have worked fine.

### CODE REVIEW

File: pkgs/top-level/lua-packages.nix [hunk 1]
```diff
@@ -64,17 +64,18 @@ in
 with self; {
```
Comment (teto) [inline pkgs/top-level/lua-packages.nix]: next on my list would be to remove this dangerous `with self;`

**Comment by teto** [comment]
thanks for taking care of this, I remember adjusting LUA_PATH to work with modules that had stuff in strange places but maybe with luarocks that was fixed. Anyway this will definitely needs a few nix-review runs to make sure it works fine/fix whatever needs to be fixed. I am

**Comment by Shados** [comment]
Odd. I just force-pushed the branch to fix an issue with my change to `getPath`, but it hasn't updated the PR...

**Comment by teto** [comment]
wrong remote maybe :p ?

**Comment by Shados** [comment]
No, I followed the branch link at the top of this PR, and the commit hashes are clearly different. What the hell? I guess I'll try rebasing to master and force-push again to see if that gets github to pick it up =/.

**Comment by Shados** [comment]
Nope, that didn't work either.

**Comment by Shados** [comment]
OK, looks like it was super delayed for some reason.

**Comment by lblasc** [comment]
Did a quick test:
```
~/nixpkgs$ nix-shell -I nixpkgs=./. -p luajit.pkgs.cjson

[nix-shell:~/nixpkgs]$ echo $LUA_PATH
/nix/store/7kmls9ynfllvdny4qdf7vbljsa3jg3cx-luajit-2.1.0-beta3-lua-cjson-2.1.0.6-1/lib/lua/5.1/?.lua;/nix/store/7kmls9ynfllvdny4qdf7vbljsa3jg3cx-luajit-2.1.0-beta3-lua-cjson-2.1.0.6-1/share/lua/5.1/?.lua;/nix/store/7kmls9ynfllvdny4qdf7vbljsa3jg3cx-luajit-2.1.0-beta3-lua-cjson-2.1.0.6-1/share/lua/5.1/?/init.lua;/nix/store/7kmls9ynfllvdny4qdf7vbljsa3jg3cx-luajit-2.1.0-beta3-lua-cjson-2.1.0.6-1/lib/lua/5.1/?/init.lua;/nix/store/kqfhcx0z02ylx38af3maj8ddfsff42l0-luajit-2.1.0-beta3/lib/lua/5.1/?.lua;/nix/store/kqfhcx0z02ylx38af3maj8ddfsff42l0-luajit-2.1.0-beta3/share/lua/5.1/?.lua;/nix/store/kqfhcx0z02ylx38af3maj8ddfsff42l0-luajit-2.1.0-beta3/share/lua/5.1/?/init.lua;/nix/store/kqfhcx0z02ylx38af3maj8ddfsff42l0-luajit-2.1.0-beta3/lib/lua/5.1/?/init.lua;/nix/store/kqfhcx0z02ylx38af3maj8ddfsff42l0-luajit-2.1.0-beta3/share/luajit-2.1.0-beta3/?.lua
```

It works like before. Great work @Shados

**Comment by lblasc** [comment]
Without this patches Lua packages in nix-shell are broken (without manual `export` of `LUA_PATH` and `LUA_CPATH`). @teto / @vcunat can you merge this?

**Review by lblasc — APPROVED** [review]

**Pull request merged**